auth-sample-shared/src/main/java/com/google/android/horologist/auth/sample/shared/TokenBundleSerializer.kt

```
@@ -16,22 +16,25 @@
 
 package com.google.android.horologist.auth.sample.shared
 
+import androidx.datastore.core.CorruptionException
 import androidx.datastore.core.Serializer
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import com.google.android.horologist.auth.sample.shared.model.TokenBundleProto.TokenBundle
+import com.google.protobuf.InvalidProtocolBufferException
 import java.io.InputStream
-import java.io.InputStreamReader
 import java.io.OutputStream
 
-public object TokenSerializer : Serializer<String> {
-    override val defaultValue: String = ""
+public object TokenBundleSerializer : Serializer<TokenBundle?> {
```

I wonder whether it's simpler just to treat a TokenBundle with all blank keys as missing.  But your call. If this works, modulo the readFrom, then it seems ok.

---

I think it's confusing to treat an object instance as "missing", it can lead to "tolerable" paths in the code where it should fail fast (e.g. crash during development as some info that should have been sent was not sent) or report (e.g. server returned empty data, app should log and report it instead of just treating as missing value).

If it had to go that way then I would add an extra boolean property to make it explicit that was marked as missing, or if explore a more complex data structure like sealed classes in kotlin. I think null serves its purpose here to mark as missing, if there are not other further implications.

If one wants to avoid nullable types in the app, then a domain model can fulfil that purpose. The proto model is part of the data layer, and a null value for it can be mapped into a "TokenBundle.Missing" class from the domain.